app/src/main/java/com/vitorpamplona/amethyst/ui/screen/ChatroomListFeedView.kt

```diff
@@ -77,19 +87,85 @@ private fun FeedLoaded(
 ) {
     val listState = rememberLazyListState()
 
-    LazyColumn(
-        contentPadding = PaddingValues(
-            top = 10.dp,
-            bottom = 10.dp
-        ),
-        state = listState
-    ) {
-        itemsIndexed(state.feed.value, key = { index, item -> if (index == 0) index else item.idHex }) { index, item ->
-            ChatroomCompose(
-                item,
-                accountViewModel = accountViewModel,
-                navController = navController
-            )
+    var markedAllAsRead by remember { mutableStateOf<Boolean>(false) }
+
+    Column {
+        TextButton(
+            modifier = Modifier.fillMaxWidth(),
+            shape = MaterialTheme.shapes.large,
+            onClick = { markedAllAsRead = true }
+        ) {
+            Text(stringResource(R.string.mark_all_as_read))
+        }
+
+        LazyColumn(
+            contentPadding = PaddingValues(
+                top = 10.dp,
+                bottom = 10.dp
+            ),
+            state = listState
+        ) {
+            itemsIndexed(state.feed.value, key = { index, item -> if (index == 0) index else item.idHex }) { index, item ->
+                ChatroomCompose(
+                    item,
+                    accountViewModel = accountViewModel,
+                    navController = navController
+                )
+            }
+        }
+    }
+
+    @Composable
+    fun markAllAsRead() {
+        val accountState by accountViewModel.accountLiveData.observeAsState()
+        val account = accountState?.account ?: return
+
+        val notificationCacheState = NotificationCache.live.observeAsState()
+        val notificationCache = notificationCacheState.value ?: return
+
+        val context = LocalContext.current.applicationContext
+
+        state.feed.value.forEach { baseNote ->
+            val noteState by baseNote.live().metadata.observeAsState()
+            val note = noteState?.note
+
+            val routeForLastRead = if (note?.event == null) {
+                return@forEach
+            } else if (note.channel != null) {
+                val channelState by note.channel!!.live.observeAsState()
+                val channel = channelState?.channel
+
+                channel?.let {
+                    "Channel/${it.idHex}"
+                }
+
+            } else {
+                val replyAuthorBase = note.mentions?.first()
+
+                var userToComposeOn = note.author!!
+
+                if (replyAuthorBase != null) {
+                    if (note.author == account.userProfile()) {
+                        userToComposeOn = replyAuthorBase
+                    }
+                }
+
+                "Room/${userToComposeOn.pubkeyHex}"
+            }
+
+            routeForLastRead?.let {
+                LaunchedEffect(key1 = notificationCache) {
```

Comment:
I'm not sure what `LaunchedEffect` does or what should be used as the key.

Reply:
It allows running a side effect suspend function tied to the composable's lifecycle. The key ensures that it only runs once when the key changes. Since we only want it to run if the notification cache changes, this is correct.